### PR TITLE
PRIC-1356 Allow paddle to download individual files

### PR DESCRIPTION
**Pull request by adrianomedeirossantos**
Jira ticket: https://deliveroo.atlassian.net/browse/PRIC-1356

> ### Allow paddle to download individual files
> 
> Currently when we use paddle to download files, you have to download the entire folder. Sometimes this can contain very large files that you don’t need. 
> 
> Add an option to only download specific files.

### CODE REVIEW

File: cli/data/get.go [hunk 1]
```diff
@@ -52,31 +53,38 @@ var getCmd = &cobra.Command{
...
-		copyPathToDestination(source, args[1])
+		copyPathToDestination(source, args[1], files)
 	},
 }
 
 func init() {
 	getCmd.Flags().StringVarP(&getBranch, "branch", "b", "master", "Branch to work on")
 	getCmd.Flags().StringVar(&getBucket, "bucket", "", "Bucket to use")
 	getCmd.Flags().StringVarP(&getCommitPath, "path", "p", "HEAD", "Path to fetch (instead of HEAD)")
+	getCmd.Flags().StringVarP(&getFiles, "files", "f", "", "A list of files to download separated by comma")
```
Comment (me) [inline cli/data/get.go]: If you make this a `StringSliceVarP`, it does the splitting for you (and allows you alternatively to repeat the flag several times)

Comment (adrianomedeirossantos) [reply cli/data/get.go]: That's really cool, the code is much cleaner now :)

File: cli/data/get.go [hunk 2]
```diff
@@ -135,21 +138,38 @@ func copy(session *session.Session, source S3Path, destination string) {
...
 }
 
-func copyToLocalFiles(s3Client *s3.S3, objects []*s3.Object, source S3Path, destination string) {
+func copyToLocalFiles(s3Client *s3.S3, objects []*s3.Object, source S3Path, destination string, files []string) {
 	var (
-		wg  = new(sync.WaitGroup)
-		sem = make(chan struct{}, s3ParallelGets)
+		wg           = new(sync.WaitGroup)
+		sem          = make(chan struct{}, s3ParallelGets)
+		downloadList = filterObjects(objects, files)
```
Comment (pedrocunha) [inline cli/data/get.go]: To me reads a bit odd that we use the word "files" everywhere but then these are more like "filepaths" or "s3Keys"

File: cli/data/get.go [hunk 3]
```diff
@@ -135,21 +138,38 @@ func copy(session *session.Session, source S3Path, destination string) {
...
 	wg.Wait()
 }
 
+func filterObjects(objects []*s3.Object, files []string) []*s3.Object {
+	var downloadList []*s3.Object
+	if len(files) == 0 {
+		return objects
+	}
+	for _, key := range objects {
+		_, file := filepath.Split(*key.Key)
```
Comment (pedrocunha) [inline cli/data/get.go]: Will this work to fetch files that are within directories?

File: cli/data/get_test.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	var (
+		file = "file.csv"
+		obj  = &s3.Object{Key: &file}
+	)
+
+	result := filterObjects([]*s3.Object{obj}, []string{})
+
+	length := len(result)
+	if length != 1 {
+		t.Errorf("Failed to filter files, got: %v, want: %v.", length, length)
```
Comment (pedrocunha) [inline cli/data/get_test.go]: Shouldn't this be "want" 0 ?

File: cli/data/get.go (hunk 3 above)
Comment (adrianomedeirossantos) [reply cli/data/get.go]: Nope, as there is no recursion to search within directories. I believe we could re-visit this point if there is any use case.

File: cli/data/get_test.go (hunk 4 above)
Comment (adrianomedeirossantos) [reply cli/data/get_test.go]: `want=1` actually. Will fix that.

File: cli/data/get.go (hunk 2 above)
Comment (adrianomedeirossantos) [reply cli/data/get.go]: IMO `files` are like a whitelist with filenames (or `s3Keys). When I read this piece of code, it sounds to me like: Hey! Filter this list of objects given this list of files that I want to download.



I guess it makes sense to call it `files` given that the flag name is `files`:

```
paddle data get --files file1.csv,file2.csv trained-model/version1 dest/path
```

But, tbh I don't have strong feelings. If we are to change the variable name, I'd also change the flag name though.

**Comment by adrianomedeirossantos** [comment]
@pedrocunha [PRIC-1356](https://deliveroo.atlassian.net/browse/PRIC-1356) was about to download files (at least that was my understanding). But I'm totally in to introduce it into the pipeline if there is any use case.

**Pull request merged**